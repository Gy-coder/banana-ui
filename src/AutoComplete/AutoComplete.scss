.g-auto-complete {
  position: relative;
  display: inline-block;
  &.block {
    display: block;
  }
  > ul {
    margin-top: 0;
    margin-bottom: 1rem;
  }
  &-suggestion {
    position: absolute;
    top: calc(100% + 8px);
    left: 0;
    z-index: 100;
    width: 100%;
    padding-left: 0;
    white-space: nowrap;
    list-style: none;
    background: #fff;
    border: 1px solid #dee2e6;
    box-shadow: 0 2px 4px 0 rgb(0 0 0 / 12%), 0 0 6px 0 rgb(0 0 0 / 4%);
  }
  &-suggestion-item {
    padding: 0.5rem 1rem;
    color: #212529;
    cursor: pointer;
    transition: color 0.15s ease-in-out, border-color 0.15s ease-in-out;
    &:hover {
      color: #1890ff !important;
    }
    &.item-highlighted {
      color: #fff !important;
      background: #1890ff !important;
    }
  }
}
