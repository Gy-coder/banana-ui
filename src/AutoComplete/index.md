---
title: AutoComplete
nav:
  title: AutoComplete
  path: /components
---

# AutoComplete

输入框自动完成功能。

# 何时使用

需要一个输入框而不是选择器。

需要输入建议/辅助提示。

和 Select 的区别是：

- AutoComplete 是一个带提示的文本输入框，用户可以自由输入，关键词是辅助输入。

- Select 是在限定的可选项中进行选择，关键词是选择。

# 代码演示

## 基础使用

<code src="./demos/basic.tsx" />

<code src="./demos/width.tsx" />

<code src="./demos/renderOption.tsx" />

<code src="./demos/async.tsx" />
