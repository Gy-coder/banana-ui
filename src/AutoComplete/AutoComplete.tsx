import React, { ReactElement, useEffect, useRef, useState } from 'react';
import classnames from 'classnames';
import { LoadingOutlined } from '@ant-design/icons';
import Input, { InputProps } from '../Input/Input';
import { useDebounce } from '../hooks/useDebounce';
import { useClickOutside } from '../hooks/useClickOutside';
import './AutoComplete.scss';

export interface AutoCompleteProps
  extends Omit<InputProps, 'onSelect' | 'onChange'> {
  fetchSuggestions: (
    str: string,
  ) => DataSourceType[] | Promise<DataSourceType[]>;
  onSelect?: (item: DataSourceType) => void;
  onChange?: (str: string) => void;
  renderOption?: (item: DataSourceType) => ReactElement;
  block?: boolean;
}

interface DataSourceObject {
  value: string;
}

export type DataSourceType<T = {}> = T & DataSourceObject;

const AutoComplete: React.FC<AutoCompleteProps> = (props) => {
  const {
    fetchSuggestions,
    onSelect,
    value,
    onChange,
    block,
    renderOption,
    ...restProps
  } = props;
  const [suggestions, setSuggestions] = useState<DataSourceType[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [highlightIndex, setHighlightIndex] = useState<number>(-1);
  const triggerSearch = useRef<boolean>(false);
  const componentRef = useRef<HTMLDivElement>(null);
  const debounceValue = useDebounce(value, 200);
  useClickOutside(componentRef, () => {
    setSuggestions([]);
  });
  useEffect(() => {
    if (debounceValue && triggerSearch.current) {
      const res = fetchSuggestions(debounceValue as string);
      if (res instanceof Promise) {
        setLoading(true);
        res.then((data) => {
          setLoading(false);
          setSuggestions(data);
        });
      } else {
        setSuggestions(res);
      }
    } else {
      setSuggestions([]);
    }
    setHighlightIndex(-1);
  }, [debounceValue]);
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value.trim();
    onChange && onChange(value);
    triggerSearch.current = true;
  };
  const handleSelect = (item: DataSourceType) => {
    onChange && onChange(item.value);
    setSuggestions([]);
    onSelect && onSelect(item);
    triggerSearch.current = false;
  };
  const highlight = (index: number) => {
    if (index < 0) index = 0;
    if (index >= suggestions.length) index = suggestions.length - 1;
    setHighlightIndex(index);
  };
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    switch (e.keyCode) {
      case 13:
        if (suggestions[highlightIndex]) {
          handleSelect(suggestions[highlightIndex]);
        }
        break;
      case 38:
        highlight(highlightIndex - 1);
        break;
      case 40:
        highlight(highlightIndex + 1);
        break;
      case 27:
        setSuggestions([]);
        break;
      default:
        break;
    }
  };
  const renderTemplate = (item: DataSourceType) => {
    return renderOption ? renderOption(item) : item.value;
  };
  const generateDropDown = () => {
    return (
      <ul className="g-auto-complete-suggestion">
        {suggestions.map((item, index) => {
          const classes = classnames('g-auto-complete-suggestion-item', {
            'item-highlighted': index === highlightIndex,
          });
          return (
            <li
              className={classes}
              key={index + Math.random()}
              onClick={() => handleSelect(item)}
            >
              {renderTemplate(item)}
            </li>
          );
        })}
      </ul>
    );
  };
  return (
    <div
      className={classnames('g-auto-complete', {
        block,
      })}
      ref={componentRef}
    >
      <Input
        value={value}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        block
        {...restProps}
      />
      {loading && (
        <ul className="g-auto-complete-suggestion">
          <LoadingOutlined className="g-suggesstion-icon" />
        </ul>
      )}
      {suggestions.length > 0 && generateDropDown()}
    </div>
  );
};

AutoComplete.defaultProps = {
  block: true,
};

export default AutoComplete;
