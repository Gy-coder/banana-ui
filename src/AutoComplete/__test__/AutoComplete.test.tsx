import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
  screen,
  waitFor,
} from '@testing-library/react';
import React, { useState } from 'react';
import '@testing-library/jest-dom';
import AutoComplete, {
  AutoCompleteProps,
  DataSourceType,
} from '../AutoComplete';

const testArray = [
  { value: 'ab', number: 11 },
  { value: 'abc', number: 1 },
  { value: 'b', number: 4 },
  { value: 'c', number: 15 },
];

const testProps: AutoCompleteProps = {
  fetchSuggestions: (query) =>
    testArray.filter((item) => item.value.includes(query)),
  onSelect: jest.fn(),
  placeholder: 'This is a placeholder',
};

const testProps2: AutoCompleteProps = {
  fetchSuggestions: (query) => {
    return testArray.filter((item) => item.value.includes(query));
  },
  onSelect: jest.fn(),
  placeholder: 'auto-completed',
  renderOption: (item: DataSourceType<{ number?: number }>) => (
    <span>
      {item.value} - {item.number}
    </span>
  ),
};

let wrapper: RenderResult, inputNode: HTMLInputElement;

const Demo: React.FC<AutoCompleteProps> = (props) => {
  const [value, setValue] = useState('');
  const onChange = (value: string) => {
    setValue(value);
  };
  return <AutoComplete value={value} onChange={onChange} {...props} />;
};

describe('test AutoComplete component', () => {
  beforeEach(() => {
    wrapper = render(<Demo {...testProps} />);
    inputNode = wrapper.getByPlaceholderText(
      'This is a placeholder',
    ) as HTMLInputElement;
  });
  it('should render correct', () => {
    cleanup();
    const { asFragment } = render(
      <AutoComplete data-testid="auto" {...testProps} />,
    );
    expect(screen.getByTestId('auto')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it('test basic AutoComplete behavior', async () => {
    fireEvent.change(inputNode, { target: { value: 'a' } });
    expect(inputNode.value).toBe('a');
    await waitFor(() => {
      expect(screen.queryByText('ab')).toBeInTheDocument();
    });
    expect(
      wrapper.container.querySelectorAll('.g-auto-complete-suggestion-item')
        .length,
    ).toBe(2);
    expect(wrapper.asFragment()).toMatchSnapshot();
    fireEvent.click(wrapper.getByText('ab'));
    expect(testProps.onSelect).toHaveBeenCalledWith({
      value: 'ab',
      number: 11,
    });
    expect(inputNode.value).toBe('ab');
    expect(wrapper.queryByText('ab')).not.toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
  it('should provide keyborad support', async () => {
    fireEvent.change(inputNode, { target: { value: 'a' } });
    await waitFor(() => {
      expect(wrapper.queryByText('ab')).toBeInTheDocument();
    });
    const firstResult = wrapper.queryByText('ab');
    const secondResult = wrapper.queryByText('abc');

    // arrow down
    fireEvent.keyDown(inputNode, { keyCode: 40 });
    expect(wrapper.queryByText('ab')).toHaveClass('item-highlighted');
    // arrow down
    fireEvent.keyDown(inputNode, { keyCode: 40 });
    expect(wrapper.queryByText('abc')).toHaveClass('item-highlighted');
    // // arrow up
    fireEvent.keyDown(inputNode, { keyCode: 38 });
    expect(wrapper.queryByText('ab')).toHaveClass('item-highlighted');
    // // press enter
    fireEvent.keyDown(inputNode, { keyCode: 13 });
    expect(testProps.onSelect).toHaveBeenCalledWith({
      value: 'ab',
      number: 11,
    });
    expect(wrapper.queryByText('ab')).not.toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
  it('click outside should hide the dropdown', async () => {
    // input change
    fireEvent.change(inputNode, { target: { value: 'a' } });
    await waitFor(() => {
      expect(wrapper.queryByText('ab')).toBeInTheDocument();
    });
    fireEvent.click(document);
    expect(wrapper.queryByText('ab')).not.toBeInTheDocument();
  });
  it('renderOption should generate the right template', async () => {
    cleanup();
    wrapper = render(<Demo {...testProps2} />);
    inputNode = wrapper.getByPlaceholderText(
      'auto-completed',
    ) as HTMLInputElement;
    fireEvent.change(inputNode, { target: { value: 'a' } });
    await waitFor(() => {
      expect(wrapper.queryByText('ab - 11')).toBeInTheDocument();
    });
    expect(wrapper.asFragment()).toMatchSnapshot();
    expect(
      wrapper.container.querySelectorAll('.g-auto-complete-suggestion-item')
        .length,
    ).toEqual(2);
  });
});
