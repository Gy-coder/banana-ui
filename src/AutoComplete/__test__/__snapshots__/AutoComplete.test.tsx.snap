// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AutoComplete component renderOption should generate the right template 1`] = `
<DocumentFragment>
  <div
    class="g-auto-complete block"
  >
    <div
      class="g-input-wrapper block"
    >
      <input
        class="g-input"
        placeholder="auto-completed"
        value="a"
      />
    </div>
    <ul
      class="g-auto-complete-suggestion"
    >
      <li
        class="g-auto-complete-suggestion-item"
      >
        <span>
          ab - 11
        </span>
      </li>
      <li
        class="g-auto-complete-suggestion-item"
      >
        <span>
          abc - 1
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`test AutoComplete component should provide keyborad support 1`] = `
<DocumentFragment>
  <div
    class="g-auto-complete block"
  >
    <div
      class="g-input-wrapper block"
    >
      <input
        class="g-input"
        placeholder="This is a placeholder"
        value="ab"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test AutoComplete component should render correct 1`] = `
<DocumentFragment>
  <div
    class="g-auto-complete block"
  >
    <div
      class="g-input-wrapper block"
    >
      <input
        class="g-input"
        data-testid="auto"
        placeholder="This is a placeholder"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`test AutoComplete component test basic AutoComplete behavior 1`] = `
<DocumentFragment>
  <div
    class="g-auto-complete block"
  >
    <div
      class="g-input-wrapper block"
    >
      <input
        class="g-input"
        placeholder="This is a placeholder"
        value="a"
      />
    </div>
    <ul
      class="g-auto-complete-suggestion"
    >
      <li
        class="g-auto-complete-suggestion-item"
      >
        ab
      </li>
      <li
        class="g-auto-complete-suggestion-item"
      >
        abc
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`test AutoComplete component test basic AutoComplete behavior 2`] = `
<DocumentFragment>
  <div
    class="g-auto-complete block"
  >
    <div
      class="g-input-wrapper block"
    >
      <input
        class="g-input"
        placeholder="This is a placeholder"
        value="ab"
      />
    </div>
  </div>
</DocumentFragment>
`;
